doc/design/migration_to_python.md

```diff
@@ -0,0 +1,479 @@
+  // `predict_target` specifies the column to store predict result.
+  // For example: "iris.predict.class"
+  optional string predict_target = 11;
+  // `original_sql` is the original statement from that this `Statement` is
+  // generated. This can be for diagnostic purpose.
+  optional string original_sql string = 100;
+}
+
+message Program {
+  // `datasource` is the connection string to the database
+  optional string datasource = 1;
+  // `statements` is a list of `Statement` to be executed as a workflow
+  repeated Statement statements = 2;
+}
+```
+
+###### Golang
+
+The golang package `ir` should fill the `Program` message according to the parsed result.
+
+#### Workflow And Platform
+
+###### Python
+
+As decribed earlier, we would reimplement `workflow` in python. 
+
+The module `execute.py` should define at least the following functions:
+
+```python
+def execute(program:sqlflow_ir_pb2.Program) -> str:
+    """
+    `execute` executes the SQLFlow `program` and returns the job id
+    """
+    pass
+  
+def fetch(workflow_id:str):
+    """
+    `fetch` gets output of the workflow of `workflow_id` in an incremental way
+    """
+    pass
+  
+def step(statement:sqlflow_ir_pb2.Statement):
+    """
+    `step` dispatches the `statement` to the `sqlflow_submitter` package
+           according to `statement.type` and the platform configured
+    """
+    pass
+  
+def main(mode:str):
+    """
+    `main` parses a `sqlflow_ir_pb2.Program` from stdin and call `execute`
+           or `step` to execute the Program. 
+           Or get a string from stdin and call `fetch` to get the result.
+    """
+    pass
+  
+if __name__ == '__main__':
+    main(sys.argv[1] if len(sys.argv) > 1 else "execute")
+```
+
+The python function `execute` executes the SQLFlow `program` and should do at least the following tasks:
+
+- Do semantic checking for the `program` using `contracts.py`. **Diagnostics are required**.
+- Call `features.py:generate_analysis_steps` to generate additional data transforming steps. 
+- Call `couler` to generate an Argo `.yaml` including all `statement` steps and the additional steps and submit the workflow to Kubernetes.
+
+The python function `fetch` gets output of an executing `Program` in a incremental way.
+
+The python function `step` executes a SQLFlow `statement` as a workflow step. `step` should do at least the following tasks:
+
+- From the `platform` package, **locate** the python module of the **configured** platform.
+  - The configure mechanism should be described in detail in the detailed design doc (the current architecture use the environment variable `SQLFlow_submitter` to config the platform)
+  - The locating mechanism should be described in detail in the detailed design doc.
+-  Call one of `train`, `predict`, `explain` ,`run`, `solve`, `evaluate` of the python module of the platform according to `statement.type`
+
+There should be at least three modules in the `platform` package:
+
+- `pai.py`
+- `alisa.py`
```

Comment:
I think we only have `alisa` and `default`?

Reply:
The current architecture supports `pai`. I propose to duplicate all the functionalities of the current architecture at the **MVP** phase.